### ISSUE TITLE
feat: add Farcaster avatar as resolver

### ISSUE THREAD

**Issue by bonustrack** [snapshot-labs/stamp#180]
Add a new resolver for avatar returning Farcaster avatar image for a specific Ethereum address. Here is an example of resolver that we did for Lens: https://github.com/snapshot-labs/stamp/blob/master/src/resolvers/lens.ts. We should have the same for Farcaster. The function take as input an Ethereum address and return a resized image or false. 

This would allow site like Snapshot and others using Stamp to return Farcaster avatars when available.

**Comment by Mr94t3z** [comment]
okay, i'll try my best to do it.

**Comment by SamuelLHuber** [comment]
The issue I see here is that Wallet Address for any farcaster user can be queried directly on hubs, but the inverse going from wallet address to find a farcaster user name is not available on Hubs.

Hubs are the farcaster protocol nodes.

So you'd need to go through all Farcaster users and find if there is any wallet connected.

There are API providers that do the lookup quickly for you and provide the Wallet Address to Farcaster user query, but they are paid. I am unsure if that would satisfy your requirements for Stamp. adding the API key to .env requirements would solve for it. anyone running would then need a API key there though which is paid.

once one has a farcaster username (fname) or id (FID) the lookup of the profile picture is free using publicly available hubs (farcaster nodes)

**Comment by bonustrack** [comment]
@SamuelLHuber Did you tried the GRPC API "GetVerification" function resolving fid based on an Ethereum address? https://docs.farcaster.xyz/reference/hubble/grpcapi/verification

**Comment by SamuelLHuber** [comment]
> @SamuelLHuber Did you tried the GRPC API "GetVerification" function https://docs.farcaster.xyz/reference/hubble/grpcapi/verification
![Screenshot 2024-03-31 at 00 27 12](https://github.com/snapshot-labs/stamp/assets/40248495/547251c3-b406-4736-b466-b6397da15aaa)


this also requires and FID

**Comment by bonustrack** [comment]
Ok then we can use the paid API

**Comment by Mr94t3z** [comment]
I use this free API to get `avatarUrl`
https://searchcaster.xyz/api/profiles?connected_address=0xcB46Bfb7315eca9ECd42D02C1AE174DA4BBFf291

**Comment by bonustrack** [comment]
Good catch @Mr94t3z

**Comment by SamuelLHuber** [comment]
> I use this free API to get `avatarUrl` https://searchcaster.xyz/api/profiles?connected_address=0xcB46Bfb7315eca9ECd42D02C1AE174DA4BBFf291

for reference. Here's the github for Searchcaster https://github.com/gskril/searchcaster

**Comment by Mr94t3z** [comment]
> I use this free API to get `avatarUrl` https://searchcaster.xyz/api/profiles?connected_address=0xcB46Bfb7315eca9ECd42D02C1AE174DA4BBFf291

then the code should be like this:
```ts
import axios from 'axios';
import { getAddress, isAddress } from '@ethersproject/address';
import { resize } from '../utils';
import { max } from '../constants.json';
import { fetchHttpImage, axiosDefaultParams } from './utils';

const API_URL = 'https://searchcaster.xyz/api/profiles';

export default async function resolve(address) {
  const formattedAddress = getAddress(address);
  if (!isAddress(formattedAddress)) return false;

  try {
    const response = await axios.get(`${API_URL}?connected_address=${formattedAddress}`);
    const profiles = response.data;

    if (profiles.length === 0) return false;

    const avatarUrl = profiles[0]?.body?.avatarUrl;
    if (!avatarUrl) return false;

    const input = await fetchHttpImage(avatarUrl);
    return await resize(input, max, max);
  } catch (error) {
    console.error('Error resolving Farcaster avatar:', error);
    return false;
  }
}
```

**Comment by bonustrack** [comment]
Great! Can you make a pull request? @Mr94t3z

**Comment by SamuelLHuber** [comment]
> Great! Can you make a pull request? @Mr94t3z

added pull request with axios default params from utils based off of code shared above by @Mr94t3z

**Comment by Mr94t3z** [comment]
> Great! Can you make a pull request? @Mr94t3z

On it - https://github.com/snapshot-labs/stamp/pull/199

**Comment by s29papi** [comment]
I made interest to this earliest, but then, to be fair to farcaster, It is both possible to get the users avatar directly from farcaster for free and without any use of an unofficial farcaster indexer such as searchcaster.
This is done by using the users custody address to fetch the users `onChainIdRegistryEvent` which has the users fid:

`curl https://nemes.farcaster.xyz:2281/v1/onChainIdRegistryEventByAddress?address=`

Then we use this fid to fetch the users data which has the avatar:

`curl https://nemes.farcaster.xyz:2281/v1/userDataByFid?fid=fid&user_data_type=1`

All this from farcaster docs:

docs: 
1. https://docs.farcaster.xyz/reference/hubble/httpapi/onchain#onchainidregistryeventbyaddress
2. https://docs.farcaster.xyz/reference/hubble/httpapi/userdata#userdata-api

I'd make a pr to now

**Comment by Mr94t3z** [comment]
> I made interest to this earliest, but then, to be fair to farcaster, It is both possible to get the users avatar directly from farcaster for free and without any use of an unofficial farcaster indexer such as searchcaster. This is done by using the users custody address (which is more reliable, than using a connected address, since not all farcaster users have connected their wallets / addresses) to fetch the users `onChainIdRegistryEvent` which has the users fid:
> 
> `curl https://nemes.farcaster.xyz:2281/v1/onChainIdRegistryEventByAddress?address=`
> 
> Then we use this fid to fetch the users data which has the avatar:
> 
> `curl https://nemes.farcaster.xyz:2281/v1/userDataByFid?fid=fid&user_data_type=1`
> 
> All this from farcaster docs:
> 
> docs:
> 
> 1. https://docs.farcaster.xyz/reference/hubble/httpapi/onchain#onchainidregistryeventbyaddress
> 2. https://docs.farcaster.xyz/reference/hubble/httpapi/userdata#userdata-api
> 
> I'd make a pr to now

I think it should be `connected_address` not `custodyAddress`.
We also can use **Neynar** to make it best practice and the official one - https://docs.neynar.com/reference/user-bulk-by-address

```
curl --request GET \
     --url 'https://api.neynar.com/v2/farcaster/user/bulk-by-address?addresses=0xcB46Bfb7315eca9ECd42D02C1AE174DA4BBFf291' \
     --header 'accept: application/json' \
     --header 'api_key: NEYNAR_API_DOCS'
```

**Comment by s29papi** [comment]
@Mr94t3z from the docs you just shared the input is first for custody address, then connected address. So it is not best practice. Every farcaster address has a custody address but not all farcaster addresses have a connected or verified address.

**Comment by Mr94t3z** [comment]
@s29papi You're right, but the user won't know what's their `custodyAddress` except if the user imports the phrase. I think DEGEN - https://www.degen.tips/ didn't use `custodyAddress` tho.

**Comment by developerfred** [comment]
@bonustrack Can I open a pull request for this activity?

**Comment by s29papi** [comment]
https://github.com/snapshot-labs/stamp/pull/202

**Comment by bonustrack** [comment]
@SamuelLHuber @Mr94t3z @developerfred @s29papi Thank you guys for your PRs but I want to highlight that this bounty has only 1 winner, I don't recommend starting coding before I confirm the person selected to do this. We will be adding new and related bounties that you might be interested into, but for the moment this issue is assigned to @SamuelLHuber. He was the first to engage in Farcaster.

**Comment by Mr94t3z** [comment]
Good luck.

**Comment by SamuelLHuber** [comment]
Regarding custody address:

For most users the custody address is just what warpcast gave them and not a address that has any onchain activity, that is why the connected address is important, I'll update the [PR](https://github.com/snapshot-labs/stamp/pull/198) with Neynar @bonustrack should I use Neynar as fallback for searchcaster or just switch over to neynar and fail when there is no neynar API Key set in environment variables?

**Comment by Mr94t3z** [comment]
> @SamuelLHuber @Mr94t3z @developerfred @s29papi Thank you guys for your PRs but I want to highlight that this bounty has only 1 winner, I don't recommend starting coding before I confirm the person selected to do this. We will be adding new and related bounties that you might be interested into, but for the moment this issue is assigned to @SamuelLHuber. He was the first to engage in Farcaster.

### **You can see who's the first one reply you on Farcaster.**

> 0x94t3z

![IMG_0074](https://github.com/snapshot-labs/stamp/assets/52822242/998729c6-9c29-4204-9199-5d0daa99ebec)

> samuellhuber

![IMG_0075](https://github.com/snapshot-labs/stamp/assets/52822242/2424ae39-8ec5-420c-a3de-cb42dc8515fb)

**Comment by s29papi** [comment]
@Mr94t3z Common man, I have deleted my replies to the bounty, and moved on. 

But I interacted with @bonustrack a day before any of you. He / she asked I sent a pm but I couldn't because I didn't see a message box on his / her profile. 

It his / her bounty and I respect his / her decision to assign it who they want. The account I used is @onchainbets_bot or something. 

Yes @bonustrack said it's @SamuelLHuber, but publicly in the comment sections, I was first to interact.
